// @iden3/js-merkletree/dist/node/cjs/lib/merkletree/circom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CircomProcessorProof = exports.CircomVerifierProof = void 0;
const hash_1 = require("../hash/hash");
class CircomVerifierProof {
    constructor(_root = hash_1.ZERO_HASH, _siblings = [], _oldKey = hash_1.ZERO_HASH, _oldValue = hash_1.ZERO_HASH, _isOld0 = false, _key = hash_1.ZERO_HASH, _value = hash_1.ZERO_HASH, _fnc = 0) {
        this.root = _root;
        this.siblings = _siblings;
        this.oldKey = _oldKey;
        this.oldValue = _oldValue;
        this.isOld0 = _isOld0;
        this.key = _key;
        this.value = _value;
        this.fnc = _fnc;
    }
}
exports.CircomVerifierProof = CircomVerifierProof;
class CircomProcessorProof {
    constructor(_oldRoot = hash_1.ZERO_HASH, _newRoot = hash_1.ZERO_HASH, _siblings = [], _oldKey = hash_1.ZERO_HASH, _oldValue = hash_1.ZERO_HASH, _newKey = hash_1.ZERO_HASH, _newValue = hash_1.ZERO_HASH, _isOld0 = false, _fnc = 0) {
        this.oldRoot = _oldRoot;
        this.newRoot = _newRoot;
        this.siblings = _siblings;
        this.oldKey = _oldKey;
        this.oldValue = _oldValue;
        this.newKey = _newKey;
        this.newValue = _newValue;
        this.isOld0 = _isOld0;
        this.fnc = _fnc;
    }
}
exports.CircomProcessorProof = CircomProcessorProof;
